Eliminate dependecy on openeo-udf package

soxofaan opened the pull request:
Addresses issue #159/EP-3578:

moves the pure UDF execution parts from [openeo-udf](https://github.com/Open-EO/openeo-udf) package (so not the server implementation aspects) to openeo Python client lib

(PR still under construction at the moment, but already available for review)

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,187 @@
+        'DataCube': XarrayDataCube,
+        'XarrayDataCube': XarrayDataCube,
+        'UdfData': UdfData
+    }
```
Improve autodetection of what is available?

soxofaan commented:
Overview of what is added/changed:

- copy  and fine-tune `openeo_udf.api.datacube/DataCube` to `openeo.udf.xarraydatacube.XarrayDataCube`: I made name more explicit to avoid confusion with existing `DataCube`.
- copy and fine-tune `openeo_udf.api.run_code` to `openeo.udf.run_code` 
- copy and fine-tune `openeo_udf.api.structured_data` to `openeo.udf.structured_data`
- copy and fine-tune `openeo_udf.api.udf_data` to `openeo.udf.udf_data`  (not ported: feature_collection_list, ml_model_list, metadata, server_context)
- copy doctests/unittest and adapted to openeo python client lib style
- port/adapt and fix some unit tests that already existed in openeo python client lib

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,187 @@
+                    c=len(data.get_datacube_list())
+                ))
+            # TODO: also support calls without user context?
+            result_cube = func(data.get_datacube_list()[0], data.user_context)
```
This code path is not fully covered with tests

jdries commented:
I believe we also had this file:
openeo-udf/src/openeo_udf/api/udf_signatures.py
which was meant as documentation, to define allowed signatures supported by 'run_code'. Is it possible that this is not yet included, or is there some alternative?

soxofaan commented:
>  openeo-udf/src/openeo_udf/api/udf_signatures.py

indeed, that is still to do (I though I did put a TODO note in `docs/udf.rst` about it, but apparently that didn't get committed). 

`udf_signatures` doesn't seem completely in sync with the actual implementation however, so that as to be tackled as well

aljacob commented:
@clausmichele can you have a look if this might break our UDF deployment for python at the moment?

clausmichele commented:
Hi Stefaan, currently the python UDF instance at Eurac is not running, but we will anyway have to update it in the future.
I think there will be no problems for us!

soxofaan commented:
latest changes:
- port `openeo_udf.api.udf_signatures` to `openeo.udf.udf_signatures` 
- added `def apply_udf_data(data: UdfData)` that was supported but not described in `udf_signatures`
- bring back `_save_DataArray_to_JSON`/`_save_DataArray_to_NetCDF` which were used apparently in openeo-geopyspark-driver
- port `openeo_udf.api.feature_collection.FeatureCollection` to `openeo.udf.feature_collection.FeatureCollection` and simplified its API

soxofaan commented on docs/udf.rst:
```diff
@@ -13,25 +13,35 @@ a piece of data as possible.
+
+.. note::
+
+    Don not confuse user-defined profunctions (often abbreviated as UDF) with
```
typo

soxofaan commented on docs/udf.rst:
```diff
@@ -40,59 +50,52 @@ To ensure that openEO understand your function, it needs to follow some rules, t
-    >>> udfstr=load_udf('../examples/udf/smooth_savitzky_golay.py')
-    >>> DataCube.execute_local_udf(udfstr, 'test_input.nc', fmt='netcdf'):
+    from openeo.udf import execute_local_udf
+    execute_local_udf(smoothing_udf, 'test_input.nc', fmt='netcdf'):
```
trailing colon ?

soxofaan commented on examples/ndvi_composite.py:
```diff
@@ -104,7 +104,7 @@ def test_debug_udf_direct_invoke():
-    from openeo.rest.conversions import datacube_from_file
-    udf_cube = datacube_from_file('masked.nc', fmt='netcdf')
+    from openeo.udf import XarrayDataCube
+    udf_cube = XarrayDataCube.from_file('masked.nc', fmt='netcdf')
```
autodetect netcdf from `.nc` extension?

soxofaan replied:
done

soxofaan commented on examples/udf/median_composite.py:
```diff
@@ -1,19 +1,16 @@
     array: xarray.DataArray = cube.get_array()
 
-    import pandas as pd
     import numpy as np
```
move numpy import up

soxofaan commented on openeo/rest/conversions.py:
```diff
@@ -1,13 +1,10 @@
 import numpy as np
 import pandas
-import xarray
+from deprecated import deprecated
```
Add this module to sphinx documentation (and use sphinx `deprecated` decorator)

soxofaan commented on openeo/rest/conversions.py:
```diff
@@ -87,235 +84,29 @@ def timeseries_json_to_pandas(timeseries: dict, index: str = "date", auto_collap
 
-    if title is not None:
-        fig.text(0.,1.,title.split('/')[-1], size=fs, va="top", ha="left",weight='bold')
+@deprecated("Use :py:method:`XarrayIO.to_json_file` instead", version="0.6.2")
```
move these up to original location

soxofaan commented on openeo/rest/datacube.py:
```diff
@@ -1446,51 +1445,10 @@ def execute(self) -> Dict:
-        
-        return None
+    @deprecated(reason="Use :py:func:`openeo.udf.run_code.execute_local_udf` instead", version="0.6.2")
+    def execute_local_udf(udf: str, datacube: Union[str, 'xarray.DataArray', 'XarrayDataCube'] = None, fmt='netcdf'):
```
add minimal doc block for API docs

soxofaan commented on openeo/udf/feature_collection.py:
```diff
@@ -0,0 +1,98 @@
+from typing import Optional, Union, Any, List
```
Add note about origin of this implementation (openeo-udf project)

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,232 @@
+import functools
```
Add note about origin of this implementation (openeo-udf project)

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,232 @@
+    }
+
+
+def apply_timeseries_example(series: Series, context: Dict) -> Series:
```
Is it necessary to have this example if we already have udf_signatures?

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,232 @@
+
+def _apply_timeseries_xarray(
+        array: xarray.DataArray,
+        callback: Callable[[Series], Series] = apply_timeseries_example,
```
why a default value for `callback`?

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,232 @@
+
+def apply_timeseries_generic(
+        udf_data: UdfData,
+        callback: Callable[[Series, dict], Series] = apply_timeseries_example
```
why default for `callback`?

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,232 @@
+
+def run_udf_code(code: str, data: UdfData) -> UdfData:
+    # TODO: current implementation uses first match directly, first check for multiple matches?
+    # TODO: more test coverage for this
```
remove TODO about more test coverage

soxofaan commented on openeo/udf/run_code.py:
```diff
@@ -0,0 +1,232 @@
+                and _annotation_is_pandas_series(sig.return_annotation)
+        ):
+            # this is a UDF that transforms pandas series
+            return apply_timeseries_generic(data, func)
```
add logging of which UDF type was found

soxofaan commented on openeo/udf/structured_data.py:
```diff
@@ -0,0 +1,36 @@
+import builtins
```
Add note about origin of this implementation (openeo-udf project)

soxofaan commented on openeo/udf/udf_data.py:
```diff
@@ -0,0 +1,138 @@
+        :param feature_collection_list: A list of VectorTile objects
+        :param structured_data_list: A list of structured data objects
+        """
+        # TODO: when None is given: assign empty list or just None to keep things simple?
```
remove all the boilerplate related to this and simplify the implementation

soxofaan commented on openeo/udf/udf_data.py:
```diff
@@ -0,0 +1,138 @@
+        """Delete all structured data entries"""
+        self._structured_data_list = None
+
+    # TODO: get rid of these getter/setter properties to make things more explicit and readable?
```
API with explicit getters and setters will be cleaner and more self-documenting: so maybe better to remove `property` stuff for now

soxofaan replied:
will keep getters and setters, but remove "deller"

soxofaan commented on openeo/udf/xarraydatacube.py:
```diff
@@ -0,0 +1,353 @@
+import collections
```
add note about origin of this module
